### DISCUSSION TITLE
Emojis containing VS16 (U+FE0F) are displayed as the text style

### DISCUSSION THREAD

**Discussion by tats-u** [google/mozc#1344] [Ideas]
## Description

Some emojis are built on the top of existing text symbols.
However, they are shown as the text style even though VS16 (U+FE0F) that shows the symbol should be displayed as the emoji style if possible.
Affected characters are annotated as [Symbol] [VS16].

https://ja.wikipedia.org/wiki/Unicode6.0%E3%81%AE%E6%90%BA%E5%B8%AF%E9%9B%BB%E8%A9%B1%E3%81%AE%E7%B5%B5%E6%96%87%E5%AD%97%E3%81%AE%E4%B8%80%E8%A6%A7

## Steps to reproduce

1. Type "まるひ"
2. Press Space

## Expected behavior

- Only "絵文字 絵文字" one should be displayed as "㊙️" (emoji style)
- "丸秘" one should be displayed as "㊙︎" (text style)

## Actual behavior
A clear and concise description of what you actually got.

Both are displayed as "㊙︎" (text style)

## Screenshots

![image](https://github.com/user-attachments/assets/14aaad97-7ce1-4180-8add-327d848d223f)

It's difficult to choose one If we want to force the emoji style.

## Version or commit-id
[e.g. Mozc-2.28.4960.100+24.11.oss or d50a8b9ae28c4fba265f734b38bc5ae392fe4d25]
You can get the version string by converting "Version" or "ばーじょん".

See below

## Environment
 - Windows 11 (Windows Sandbox) + Mozc-2.30.5544.100+24.11.oss
 - (Compare) Windows 11 (Windows Sandbox) + GoogleJapaneseInput-2.30.5448.100+24.11.9
 - (Compare) Windows 11 (bare metal) + GoogleJapaneseInput-2.30.5448.100+24.11.9 (Office is installed)

## Investigations
 - Whether this issue happens on both Ibus and other IMF (e.g. Fcitx, uim).
   + I haven't tried Linux
 - Whether this issue happens on other IMEs (e.g. Anthy, SKK).
   + I haven't tried Linux
 - What applications this issue happens on (e.g. Chromium, gedit).
   + Notepad
 - What applications this issue does not happen on (e.g. Chromium, gedit).
   + not-found
 - (optional) What versions or commit-ids this issue did not happen on
   (e.g. Mozc-2.28.4960.100+24.11.oss).
   not-found

## Additional context

Choose "絵文字 絵文字" basically (similar ones that don't contain "絵文字" don't contain VS16):

- ０ １ ... ９ | 0️⃣ ... 9️⃣
- ＃ | #️⃣
- ちょさくけん | ©️
- とうろくしょうひょう | ®️
- ！！ | ‼️
- ！？  | ⁉️
- さゆう | ↔️
- じょうげ | ↕️
- やじるし (絵文字 〇〇) | ↔️ ... ⬇️
- しかく | ▪️ ... ⬜️
- (はれ) (both emoji in Windows Sandbox) |☀️
- (くもり) (both emoji in Windows Sandbox) | ☁️
- (おんせん)  (both emoji in Windows Sandbox) | ♨️
- (でんわ) (both emoji in Windows Sandbox) | ☎️
- いおりてん | 〽️
- まる | Ⓜ️ ... ⭕️
- ほし | ⭐️
- (けいこく) (both emoji in Windows Sandbox) | ⚠️
- まるひ | ㊙️

I don't think the above list covers all.

🃏️ and 🀄️ have already been displayed as emoji without VS16. It's probably because a font supporting such characters is missing.

"丸秘" doesn't contain VS15 (U+FE0E) or VS16.

**Comment by tats-u** [comment]
https://unicode.org/Public/emoji/latest/emoji-sequences.txt
https://unicode.org/Public/emoji/latest/emoji-test.txt
https://www.unicode.org/emoji/charts/emoji-style.txt

**Comment by yisibl** [comment]
Emoji rendering is indeed complex, and Chrome 131 started with a lot of improvements, but there are still some issues, you can try it in the latest version of Chrome Canary.

https://groups.google.com/a/chromium.org/g/blink-dev/c/i5AeH4ZE2RY/m/ZkTWLj93AwAJ

![Image](https://github.com/user-attachments/assets/6e688e6d-7b69-4008-9dba-421dbe12e24f)


```html
data:text/html;charset=UTF-8,<!doctype html>
<style>
  .test { font-size: 38px; margin-top: 6px; font-family: arial, sans-serif;}
  .bug1 { font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Noto Sans",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji"; }
  .bug2 { font-family: -apple-system, BlinkMacSystemFont, "Apple Color Emoji", "Segoe UI Emoji"; }
  .bug3 { font-family: sans-serif, "Apple Color Emoji", "Segoe UI Emoji"; }
  .bug4 { font-family: system-ui, "Apple Color Emoji", "Segoe UI Emoji"; }
</style>
<span class="test bug1">㊙️</span>
<span class="test bug2">㊙️</span>
<span class="test bug3">㊙️</span>
<span class="test bug4">㊙️</span>
<p>Expected:</p>
<span class="test">㊙️</span>
```

**Comment by tats-u** [comment]
Firefox (Windows):

![Image](https://github.com/user-attachments/assets/3a4a4c19-b214-4e88-aaf0-465406dfc1c0)

Chrome (Windows):

![Image](https://github.com/user-attachments/assets/0c305636-ebc7-4e0e-b047-5d4f7632cfe5)

Your example always contains VS-16.
I believe it's just a Chrome's bug. Chrome should respect VS-16 even when non-emoji fonts come first.
This bug applies to Qt GUI in Mozc, too.
However, I think Mozc doesn't use WebView.

**Comment by yisibl** [comment]
Thanks for the screenshot, what is the Chrome version?

**Comment by tats-u** [comment]
129.0.6668.100 & 129.0.6668.101

**Comment by yukawa** [comment]
I'm converting this to a discussion for now as this would compete with other feature requests in terms of engineering resource allocation.

Perhaps a simple solution would be using [Qt-based renderer](https://github.com/google/mozc/tree/master/src/renderer/qt), which we currently use only in Linux. If someone succeeds to use Qt-based renderer on Windows and complete its evaluation, I'm definitely interested in the result.
